### PR TITLE
Add [NotNull] attribute on the Should() method for object assertions

### LINKED ISSUE #1115: Support c# 8 nullable reference
Just found this https://github.com/dotnet/standard/pull/1359 .
Maybe this lib can do similar things for c# 8 nullable reference.

### PR DESCRIPTION
**Pull request by 0xced**
Compelling example:

```csharp
[Theory]
[InlineData(true)]
[InlineData(false)]
public void Test1(bool currentUser)
{
    IPrincipal principal = currentUser ? new WindowsPrincipal(WindowsIdentity.GetCurrent()) : new ClaimsPrincipal();
    IIdentity? identity = principal.Identity;
    
    identity.Should().NotBeOfType<GenericIdentity>();
    identity.IsAuthenticated.Should().BeFalse();
}
```

Fixes #1115

### CODE REVIEW

File: Src/FluentAssertions/AssertionExtensions.cs [hunk 1]
```diff
@@ -277,7 +277,7 @@ private static void ForceEnumeration<T>(T subject, Func<T, IEnumerable> enumerab
     /// current <see cref="object"/>.
     /// </summary>
     [Pure]
-    public static ObjectAssertions Should(this object actualValue)
+    public static ObjectAssertions Should([System.Diagnostics.CodeAnalysis.NotNull] this object actualValue)
```
Comment (dennisdoomen) [inline Src/FluentAssertions/AssertionExtensions.cs:280]: Why not add the attribute to _all_ `Should()` methods?

**Comment by IT-VBFK** [comment]
Isn't this something that pretends what isn't there?

In such a case we can get rid of all `null` checks inside in the assertions?

**Comment by dennisdoomen** [comment]
> Isn't this something that pretends what isn't there?
> 
> In such a case we can get rid of all `null` checks inside in the assertions?

I'm not entirely sure what you are saying? If you mean that we're kind of pretending that the subject is never `null`, that's true. We're relying on the assertion method itself to take care of that.

**Comment by IT-VBFK** [comment]
Ok, then this is just to silence the IDE, right?

**Comment by lg2de** [comment]
> Isn't this something that pretends what isn't there?
> 
> In such a case we can get rid of all `null` checks inside in the assertions?

Newer compiler version use such attributes to create warnings. Older version does not.
And: It is a warning which the user can ignore or even disable.
So, the `Should` method cannot **ensure** non-nullability with the attribute.

**Comment by IT-VBFK** [comment]
> So, the Should method cannot ensure non-nullability with the attribute.

Ahh.. thanks for clarification :)

File: Src/FluentAssertions/AssertionExtensions.cs (hunk 1 above)
Comment (0xced) [reply Src/FluentAssertions/AssertionExtensions.cs:280]: That's a very good question! I think we should add the [NotNull] attribute on all `Should` methods that return a `ReferenceTypeAssertions<,>`. So I started to write a test and noticed that I could not assert `parameterInfo.IsDecoratedWith<NotNullAttribute>`, hence #2385. 😄

**Comment by 0xced** [comment]
I have rebased this pull request on top of #2385 because the new specs (`Should_methods_returning_reference_type_assertions_are_annotated_with_not_null_attribute` and `Should_methods_not_returning_reference_type_assertions_are_not_annotated_with_not_null_attribute`) require the new assertions on `ParameterInfo`.

File: Src/FluentAssertions/AssertionExtensions.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/AssertionExtensions.cs]: > So I started to write a test and noticed that I could not assert parameterInfo.IsDecoratedWith<NotNullAttribute>, hence https://github.com/fluentassertions/fluentassertions/pull/2385. 😄

I don't think we need to have tests for that.

Comment (0xced) [reply Src/FluentAssertions/AssertionExtensions.cs]: Well, at least it was very interesting to dive into all the assertion types, I learnt a lot. And the test is now written anyway, alongside other tests on the `Should` methods. But this pull request could also be merged without the tests added in `AssertionExtensionsSpecs` and thus without requiring #2385 to also be merged.

**Comment by 0xced** [comment]
I just rebased this pull request on the `develop` branch and make it **not** depend on #2385 at all.

The assertion is now written like this:

```csharp
parameter.GetCustomAttribute<NotNullAttribute>().Should().NotBeNull();
```

instead of this (which would require #2385 to be merged first):
```csharp
parameter.Should().BeDecoratedWith<NotNullAttribute>();
```

**Comment by zhangpengchen** [comment]
Hi @dennisdoomen ,

Would you consider releasing this feature soon so we can finally avoid workarounds by mixing the native assertions and fluent assertions in the tests.

**Review by dennisdoomen — APPROVED** [review]

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Please also update the `releases.md`

**Comment by 0xced** [comment]
I just added an entry in the release notes under improvements in 6ac05e84f9e7aeb8df561b3e997db04f4e488fd5.

**Review by dennisdoomen — APPROVED** [review]
Awesome

**Review by jnyrup — APPROVED** [review]
Noticed that `[NotNull]` is not added to these overloads of `Should`, but guess it doesn't really matter that much.

```
AssertionExtensions.cs:168:    public static ExecutionTimeAssertions Should(this ExecutionTime executionTime)
AssertionExtensions.cs:753:    public static TypeSelectorAssertions Should(this TypeSelector typeSelector)
AssertionExtensions.cs:788:    public static MethodInfoSelectorAssertions Should(this MethodInfoSelector methodSelector)
AssertionExtensions.cs:813:    public static PropertyInfoSelectorAssertions Should(this PropertyInfoSelector propertyInfoSelector)
AssertionExtensions.cs:865:    public static TaskCompletionSourceAssertions<T> Should<T>(this TaskCompletionSource<T> tcs)
AssertionExtensions.cs:894:    public static TaskCompletionSourceAssertions Should(this TaskCompletionSource tcs)
```

**Comment by thj-dk** [comment]
When is this planned to be released?

**Comment by ITaluone** [comment]
I think there is no specified date for now :)

I assume, that at least this ticket have to be completely addressed: https://github.com/fluentassertions/fluentassertions/issues/1677

**Comment by dennisdoomen** [comment]
@ITaluone is correct

**Comment by dggsax** [comment]
Is it possible to release it in an alpha release at the very least? Or are there items in #1677 that you need assistance with that could help speed this along. I very much look forward to taking advantage of the changes in this PR, thanks!

**Comment by dennisdoomen** [comment]
> Is it possible to release it in an alpha release at the very leas

I just tagged alpha 4. Since it's a pre-release, all bets are off.

**Comment by 0xced** [comment]
Oopsie. 😖

https://github.com/fluentassertions/fluentassertions/actions/runs/9291245214/job/25569332022#step:5:551

> [ERR] Push: error: Response status code does not indicate success: 403 (The specified API key is invalid, has expired, or does not have permission to access the specified package.).

**Comment by IT-VBFK** [comment]
Why not just compiling the lates develop yourself?

- Clone this repo
- run `build.ps1` or `build.sh` and you have your "alpha version" 🤷‍♂️

**Comment by dennisdoomen** [comment]
Fixed the API key and pushed again.

**Pull request merged**